core/async/src/test_loop/futures.rs

```diff
@@ -1,223 +1,107 @@
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_tuple("Task").field(&self.description).finish()
-    }
-}
+use super::data::TestLoopData;
+use super::DelaySender;
 
-/// Drives any Arc<TestLoopTask> events (futures spawned by our implementation
-/// of FutureSpawner) that are remaining in the loop.
-pub fn drive_futures() -> LoopEventHandler<(), Arc<TestLoopTask>> {
-    LoopEventHandler::new_simple(|task: Arc<TestLoopTask>, _| {
-        // The following is copied from the Rust async book.
-        // Take the future, and if it has not yet completed (is still Some),
-        // poll it in an attempt to complete it.
-        let mut future_slot = task.future.lock().unwrap();
-        if let Some(mut future) = future_slot.take() {
-            let waker = waker_ref(&task);
-            let context = &mut Context::from_waker(&*waker);
-            if future.as_mut().poll(context).is_pending() {
-                // We're still not done processing the future, so put it
-                // back in its task to be run again in the future.
-                *future_slot = Some(future);
-            }
-        }
-    })
-}
+/// Support for futures in TestLoop.
+///
+/// There are two key features this file provides for TestLoop:
+///
+///   1. A general way to spawn futures and have the TestLoop drive the futures.
+///      To support this, pass test_loop.future_spawner() the &dyn FutureSpawner
+///      to any component that needs to spawn futures.
+///
+///      This causes any futures spawned during the test to end up as an callback in the
+///      test loop. The event will eventually be executed by the drive_futures function,
+///      which will drive the future until it is either suspended or completed. If suspended,
+///      then the waker of the future (called when the future is ready to resume) will place
+///      the event back into the test loop to be executed again.
+///
+///   2. A way to send a message to the TestLoop and expect a response as a
+///      future, which will resolve whenever the TestLoop handles the message.
+///      To support this, use MessageWithCallback<Request, Response> as the
+///      event type, and in the handler, call (event.responder)(result)
+///      (possibly asynchronously) to complete the future.
+///
+///      This is needed to support the AsyncSender interface, which is required
+///      by some components as they expect a response to each message. The way
+///      this is implemented is by implementing a conversion from
+///      DelaySender<MessageWithCallback<Request, Response>> to
+///      AsyncSender<Request, Response>.
 
-/// A DelaySender<Arc<TestLoopTask>> is a FutureSpawner that can be used to
+/// A DelaySender is a FutureSpawner that can be used to
 /// spawn futures into the test loop. We give it a convenient alias.
-pub type TestLoopFutureSpawner = DelaySender<Arc<TestLoopTask>>;
+pub type TestLoopFututeSpawner = DelaySender;
```

Comment:
it seems like there is no point in such type alias since you don't make the type any more specific, so I suggest to just inline that directly as `DelaySender`